caddyhttp/proxy/proxy.go

```diff
@@ -231,6 +224,16 @@ func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
 		if !keepRetrying() {
 			break
 		}
+
+		// Before we retry the request we have to make sure
+		// that the body is rewound to it's beginning.
+		//
+		// NOTE:
+		//   Reaching this point implies that keepRetrying() above returned true,
+		//   which in turn implies that body must be a *bufferedBody.
+		if err := body.(*bufferedBody).rewind(); err != nil {
```

Comment:
_Juuuuust_ to be safe, maybe this code should have one extra check. If the type assertion fails right now, that will lead to a panic (I know, it shouldn't fail, but I suspect we lack tests to enforce this right now). Either a test should be written -- this particular case might be tedious to test for, I'm not sure -- or try this, which won't panic:

```go
if bb, ok := body.(*bufferedBody); ok {
    err := bb.rewind()
    if err != nil {
        // ...
    }
}
```